gutenberg/blocks/ts/src/components/drag-and-drop-select-list.tsx

```diff
@@ -0,0 +1,187 @@
+   */
+  public componentDidMount = async () => {
+    const selected: any[] = [];
+    const items: string[] = [];
+
+    for (let i = 0; i < this.props.value.length; i++) {
+      selected.push(this.props.value[i]);
+    }
+
+    this.props.forcedItems.forEach(item => {
+      if (!selected.includes(item)) {
+        selected.push(item);
+      }
+    });
+
+    this.props.optionalItems.forEach(item => {
+      if (!selected.includes(item)) {
+        items.push(item);
+      }
+    });
+
+    this.setState({
+      items,
+      selected
+    });
+  };
+
+  draggedItem: any;
+  draggedIdx: any;
+
+  onDragStart = (e: any, index: number) => {
+    this.draggedItem = this.state.selected[index];
+    e.dataTransfer.effectAllowed = 'move';
+    e.dataTransfer.setData('text/html', e.target.parentNode);
+    e.dataTransfer.setDragImage(e.target.parentNode, 20, 20);
+  };
+
+  onDragOver = (index: number) => {
+    const draggedOverItem = this.state.selected[index];
+
+    if (this.draggedItem === draggedOverItem) {
+      return;
+    }
+    let selected = this.state.selected.filter(item => item !== this.draggedItem);
+    selected.splice(index, 0, this.draggedItem);
+    this.setState({ selected });
+    this.triggerSelectedChange(selected);
+  };
+
+  onDragEnd = () => {
+    this.draggedIdx = null;
+  };
+
+  render() {
+    return (
+      <div className='App'>
+        <main>
+          <ul>
+            {this.renderSelectedItems()}
+            {this.renderItems()}
+          </ul>
+        </main>
+      </div>
+    );
+  }
+
+  /**
+   * Renders selected items
+   */
+  private renderSelectedItems = () => {
+    const { CheckboxControl } = wp.components;
+
+    return (
+      <div>
+        {this.state.selected.map((selectedItem, idx) => (
+          <li key={selectedItem} onDragOver={() => this.onDragOver(idx)}>
+            <div className='drag-item' draggable onDragStart={e => this.onDragStart(e, idx)} onDragEnd={this.onDragEnd}>
+              <span className='menu-icon'>{MenuIcon}</span>
+              <CheckboxControl
+                checked={true}
+                label={`${idx + 1}. ${__(selectedItem, 'linkedevents')}`}
+                key={selectedItem}
+                onChange={() => this.onSelectedItemChange(selectedItem)}
+              />
+
+            </div>
+          </li>
+        ))}
+      </div>
+    );
+  };
+
+  /**
+   * Renders match items
+   */
+  private renderItems = () => {
+    const { CheckboxControl } = wp.components;
+
+    return this.state.items.map(item => {
+      return <div className='unselected-item'>{<CheckboxControl checked={false} label={__(item, 'linkedevents')} key={item} onChange={() => this.onMatchItemChange(item)} />}</div>;
+    });
+  };
+
+  /**
+   * Event handler for selected item change
+   *
+   * @param removedItem item removed from selected list
+   */
+  private onSelectedItemChange = (removedItem: string) => {
+    if (!this.props.forcedItems.includes(removedItem)) {
```

Comment:
Could you explain how this method is supposed to work?

Reply:
Method checks if the item that is trying to be deselected, is forcedItem -> therefore cannot be deselected. If not forced, it removes item from selected list and adds the item back to selectable list.